pkg/rbd/controllerserver.go

```diff
@@ -433,23 +436,66 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 		}, nil
 	}
 
+	// this will be used to unreserve snapshot if we failed to get snapshot
+	// information from the ceph backend
+	updatedRbdImage := ""
+	// update the rbd image name to clone from parent volume
+	rbdSnap.RbdImageName = rbdVol.RbdImageName
 	err = reserveSnap(rbdSnap, cr)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	defer func() {
 		if err != nil {
+			if updatedRbdImage != "" {
+				rbdSnap.RbdImageName = updatedRbdImage
+			}
 			errDefer := undoSnapReservation(rbdSnap, cr)
 			if errDefer != nil {
 				klog.Warningf("failed undoing reservation of snapshot: %s %v", req.GetName(), errDefer)
 			}
 		}
 	}()
 
-	err = cs.doSnapshot(rbdSnap, cr)
+	// generate temp snap struct and vol struct to create a temprory snapshot
+	// and to create a new volume which will be used to take new snapshot
+	tmpSnap := &rbdSnapshot{}
+	copySnapStruct(tmpSnap, rbdSnap)
+
+	// using request name as the temp snapshot request name. this will help us to
+	// identify if the temp snap is already created.
+	// this will look like `csi-tmp-snap-1-PVC-XXXX-XXX-XXXX`
+	tmpSnap.RequestName = snapJournal.GetTmpNamePrefix(req.GetName(), true, true)
+	// generate tempVolume struct
+	tmpVol := generateVolFromSnap(rbdSnap)
+	tmpVol.ImageFeatures = "layering,deep-flatten"
+	// using request name as the volume request name. this will help us to
+	// identify if the temp snap is already created.
+	// this will look like `csi-tmp-clone-PVC-XXXX-XXX-XXXX`
+	tmpVol.RequestName = volJournal.GetTmpNamePrefix(req.GetName(), true, false)
+	err = cs.createSnapFromClone(tmpVol, tmpSnap, cr)
 	if err != nil {
 		return nil, err
 	}
+	// update the parent image to create snapshot
+	rbdSnap.RbdImageName = tmpVol.RbdImageName
+	err = createSnapshot(rbdSnap, cr)
+	if err != nil {
+		return nil, err
+	}
+	err = updateReservedSnap(rbdSnap, cr)
```

Comment:
Why do we need to update the metadata? Shouldn't we know all the finalized names before we start the process?